promql/testdata/operators.test

```diff
@@ -398,7 +398,7 @@ load 5m
 
 # On with no labels, for metrics with no common labels.
 eval instant at 5m random + on() metricA
-  {} 5
+  {foo="bar"} 5
```

Comment:
This isn't right, the fixed logic should only apply to comparison operators.

Reply:
Hmm, though #5326 was only about comparison operators I implicitly ended up assuming that it should behave the same for arithmetic operators (though the current behavior feels the most wrong in the filtering case). Unfortunately for arithmetic operators we also never specified the behavior properly in the docs at https://prometheus.io/docs/prometheus/latest/querying/operators/#arithmetic-binary-operators.

Naively I would expect the matching clause (`on()` / `ignoring()`) to be orthogonal to what labels are included in the output? I can see how it is slightly arbitrary to take the LHS vs. the RHS labels then, but then at least it'd be consistent with filtering ops?

Reply:
> Naively I would expect the matching clause (on() / ignoring()) to be orthogonal to what labels are included in the output?

For me the matching indicates the output labels, as that's the group you're working with.

Reply:
Interesting. That is a different viewpoint from where I'm coming from, which is more like treating `on()` like the `WHERE` in a SQL join and that being independent of the `SELECT *`.

Reply:
If you want that behaviour, there's group_left.

Reply:
True. But should it not be the default? I think of using `group_left` usually only in situations where there's more actual dimensionality/cardinality on the left side.

Reply:
That'd be a pretty massive breaking change.

Reply:
Hm yes. It's a breaking change also for filtering, but that's smaller, so we just accept it there? Just out of interest then, would we want to make this change if breaking-ness wouldn't be an issue (in Prometheus 3.0, say...)?

Reply:
For filtering it's a lot smaller, and in line with the intended behaviour. That's not so for general math.

Reply:
@brian-brazil I added a fixup commit to change arithmetic binops back to the old behavior. One thing I'm wondering about: if we have `CardOneToMany` for a compop, then the LHS/RHS are swapped and thus we choose the RHS labels for the output. But then the `group_x(...)` which always adds labels from the RHS (which is always the "many" side) is basically a no-op in that case, since all the RHS labels are already present. That's fine, right?

So cases like this, where the "one" side coincides with the actual initial LHS:

```
foo / on(x) group_right(y) bar
```

(`bar` would become the LHS, `foo` the RHS)

Reply:
I'd expect that to (usually) return an error as the `foo` is the labels that are returned, so will likely have duplicate series. I presume you meant to use a comparison operator in that example.

Reply:
Sorry, yes, I meant to use a comparison operator. Good point about the error.

For:

```
sum without(cpu) (node_cpu_seconds_total) == on(mode) group_right(job) node_cpu_seconds_total
```

I get `Error executing query: multiple matches for labels: grouping labels must ensure unique matches` with my change, while it works on `master`: http://demo.robustperception.io:9090/graph?g0.range_input=1h&g0.expr=sum%20without(cpu)%20(node_cpu_seconds_total)%20%3D%3D%20on(mode)%20group_right(job)%20node_cpu_seconds_total&g0.tab=1

It makes sense of course since now we don't add the additional `cpu` label from the right anymore. Maybe for `OneToMany` with comparison operators we still should, to not break things too much? Gah, this is getting complicated...

Reply:
> Gah, this is getting complicated...

Yeah, I've previously wondered if we should remove group_right. I'm now wondering if we want to change this logic at all.

Reply:
Yeah. Actually producing an error where there was none before is more breaking than I anticipated.

Of course one could say that in the case of one-to-many / many-to-one, we always choose the labels from the "many" side as the output, even if it happens to be on the RHS. So normally comp ops would always return the LHS labels, except for the one-to-many case.

Either that, or keep everything as is, I think.

Reply:
That'd be an exception on an exception, which is getting weird for something that should very rarely come up.

Our current behaviour is defendable on consistency grounds, so I'm tending towards leaving as-is currently.

Reply:
Yeah. In a greenfield world I'd lean towards just having this case produce an error, but given the breakingness of it I'll close it for now. Good to have explored this though.

Reply:
We should clarify the docs though.

Reply:
Here's an attempt: https://github.com/prometheus/prometheus/pull/5512